In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib notebook

In [5]:
UtorontoData = pd.read_csv("DataSetUtoronto.csv")

In [6]:
UtorontoData.head(5)

O-Ring failure  Launch temperature  Leak-check pressure
0               0                  66                   50
1               1                  70                   50
2               0                  69                   50
3               0                  68                   50
4               0                  67                   50

In [7]:
y = UtorontoData['O-Ring failure']
X = UtorontoData[['Launch temperature','Leak-check pressure']]

In [10]:
from sklearn.linear_model import LogisticRegression
lm = LogisticRegression()
lm.fit(X,y)
lm.coef_

array([[-0.06258052,  0.01598468]])

We learn that for every degree increase in tempretaure the odds of O-Ring Failure goes down by 6 percent. Also, for every unit of leak-check pressure the odds of O-Ring Failure goes up by 1.5%. We shall look up p-values to ensure reliability of our results. 

In [11]:
from sklearn.metrics import confusion_matrix
y_hat = lm.predict(X)
confusion_matrix(y_hat, y)

array([[17,  3],
       [ 0,  3]])

The confusion matrix suggests that we only misclassify 3 instances.

Since our analysis is only on a small sample and we have two input variables, it is quite likely that we over fit. A better measure of accuracy is Cross Validation> (we could assign some observations as test but since we only had few observations we do not do that). For fewer observations cross-validation is a better option.

In [13]:
from sklearn.cross_validation import cross_val_score
cross_val_score(lm,X,y,cv=10).mean()

//anaconda/lib/python2.7/site-packages/sklearn/cross_validation.py:516: Warning: The least populated class in y has only 6 members, which is too few. The minimum number of labels for any class cannot be less than n_folds=10.
  % (min_labels, self.n_folds)), Warning)


0.83333333333333326

our precision is 83.33% - our Cross Validation error is 16.66%. Conisdering low number of failures in our dataset that is not too good. 

#### We now, explore ROC and AUC of the estimates. 

In [17]:
from sklearn.metrics import roc_curve, auc,roc_auc_score
y_hat_probability = lm.predict_proba(X).T[1]  #T[1] returns probability of belonging to class
print(roc_auc_score(y))
#vals = roc_curve(y_hat_probability,y)
#Roc_DataFrame = pd.DataFrame({'False_Positive_Rate':vals[0],'True_Positive_Rate':vals[1]})
#Roc_DataFrame.plot(x = 'False_Positive_Rate' , y = 'True_Positive_Rate' ) #beautiful